### PR TITLE
Backport Apache Beam PR/934

### PR DESCRIPTION
**Pull request by lukecwik**
R: @peihe 
CC: @dhalperi

### CODE REVIEW

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 1]
```diff
@@ -1100,30 +1102,25 @@ private TextWriteOperation(TextSink<T> sink, Coder<T> coder, String header, Stri
...
-
-      public TextWriter(FileBasedWriteOperation<T> writeOperation, Coder<T> coder, String header,
-                        String footer) {
+      public TextWriter(FileBasedWriteOperation<T> writeOperation, Coder<T> coder,
+          @Nullable String header, @Nullable String footer) {
         super(writeOperation);
         this.header = header;
         this.footer = footer;
         this.mimeType = MimeTypes.TEXT;
         this.coder = coder;
```
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: checkNotNull

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 2]
```diff
@@ -1077,10 +1078,11 @@ private boolean tryToEnsureNumberOfBytesInBuffer(int minCapacity) throws IOExcep
...
-      private final String header;
-      private final String footer;
+      @Nullable private final String header;
+      @Nullable private final String footer;
 
-      private TextWriteOperation(TextSink<T> sink, Coder<T> coder, String header, String footer) {
+      private TextWriteOperation(TextSink<T> sink, Coder<T> coder,
+          @Nullable String header, @Nullable String footer) {
         super(sink);
         this.coder = coder;
```
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: checkNotNull

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 3]
```diff
@@ -1046,19 +1046,20 @@ private boolean tryToEnsureNumberOfBytesInBuffer(int minCapacity) throws IOExcep
...
+    @Nullable private final String footer;
 
     @VisibleForTesting
     TextSink(
-        String baseOutputFilename, String extension, String header, String footer,
+        String baseOutputFilename, String extension,
+        @Nullable String header, @Nullable String footer,
         String fileNameTemplate, Coder<T> coder) {
       super(baseOutputFilename, extension, fileNameTemplate);
       this.coder = coder;
```
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: checkNotNull

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 4]
```diff
@@ -534,9 +532,9 @@ private Read() {}
...
 
-      private Bound(String name, String filenamePrefix, String filenameSuffix, String header,
-                    String footer, Coder<T> coder, int numShards, String shardTemplate,
-                    boolean validate) {
+      private Bound(String name, String filenamePrefix, String filenameSuffix,
+          @Nullable String header, @Nullable String footer, Coder<T> coder, int numShards,
+          String shardTemplate, boolean validate) {
         super(name);
         this.header = header;
         this.footer = footer;
```
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: checkNotNull for the rest

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java (hunk 1 above)
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: Was going to leave stuff as is for code that wasn't affected

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java (hunk 2 above)
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: Was going to leave stuff as is for code that wasn't affected

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java (hunk 3 above)
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: Was going to leave stuff as is for code that wasn't affected

**Comment by peihe** [comment]
Few comments, and LGTM

@dhalperi suggests no merges during 1.7 release?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java (hunk 4 above)
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: Was going to leave stuff as is for code that wasn't affected

**Comment by lukecwik** [comment]
Wasn't going to merge till either Dan hit the merge button or got confirmation from Dan that 1.7 was cut.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 5]
```diff
@@ -1133,14 +1130,14 @@ protected void prepareWrite(WritableByteChannel channel) throws Exception {
         out = Channels.newOutputStream(channel);
       }
 
-     @Override
+      @Override
       protected void writeHeader() throws Exception {
-        writeLine(header);
+       writeIfNotNull(header);
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: missing space?

Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: Fixed

**Pull request merged**